File: C(++)/022. Generate Parentheses.cpp
/*
022. Generate Parentheses
https://leetcode.com/problems/generate-parentheses/

cpp 325 ms
*/

/*
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
*/

class Solution {
public:
    set<string> st;
    int depth;
    void step(string s,int p)
    {
        if (depth==p)
        {
            st.insert(s);
            return;
        }
        int len=s.length(),le2=len/2;
        for (int i=0;i<=len;i++)
        {
            if (i-2>0&&s[i-2]=='('&&s[i-1]==')')
                continue;
            string t=s;
            t.insert(i,"()");
            step(t,p+1);
        }
    }
    vector<string> generateParenthesis(int n) {
        st.clear();
        depth=n;
        step(string(""),0);
        vector<string> ret;
        set<string>::iterator it=st.begin();
        for (;it!=st.end();it++)
            ret.push_back(*it);
        return ret;
    }
};

File: Java/022. Generate Parentheses.java
/*
022. Generate Parentheses
https://leetcode.com/problems/generate-parentheses/

java 74 ms
*/

public class Solution {
	public HashSet<String> hret;
	public List<String> ret;
	public int _n;
	void move(char str[],int len,int p,int n)
	{
		for (int i=len-1;i>=p;i--)
			str[i+n]=str[i];
	}
	void dfs(int depth,char str[])
	{
		if (depth==_n)
		{
			hret.add(String.valueOf(str));
			return;
		}
		int len=depth*2;
		for (int i=0;i<len;i++)
		{
			char newstr[]=Arrays.copyOf(str,_n*2);
			move(newstr,len,i,2);
			newstr[i]='(';
			newstr[i+1]=')';
			dfs(depth+1,newstr);
		}
	}
	public List<String> generateParenthesis(int n) {
		hret=new HashSet<String>();
		ret=new ArrayList<String>();
		_n=n;
		if (n==0) return ret;
		
		char str[]=new char[n*2];
		str[0]='(';
		str[1]=')';
		dfs(1,str);
		ret.addAll(hret);
		return ret;
	}
}
